url9582d546a9_twentylemon_sudoku_generator_tgz-pJ8-sudoku_Top95J8/src/sudoku/Main.java

```diff
@@ -0,0 +1,123 @@
+     * @param output where to save the problems
+     */
+    public static void writeFile(List<SudokuBoard> problems, String output){
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(output))){
+            for (SudokuBoard board : problems){
+                writer.write(board + "\n\n");
+            }
+            writer.close();
+            System.out.println("output file created: " + output);
+        } catch (IOException ex){
+            System.err.println("FAILED TO WRITE OUTPUT FILE");
+        }
+    }
+
+    /**
+     * Runs the program.
+     * @param generator what to use to generate the sudoku problems
+     * @param numToCreate how many problems to create
+     * @param output where to store the problems
+     * @param verbose dump problems to screen if on
+     */
+    public static void run(SudokuGenerator generator, int numToCreate, String output, final boolean verbose){
+        List<SudokuBoard> problems = new ArrayList<>(numToCreate);
+        int minClues = Integer.MAX_VALUE;
+        SudokuBoard minBoard = null;
+        int totalClues = 0;
+        long totalTime = 0;
+        //maxTime is always the first board generated due to JIT it seems
+        //long maxTime = Long.MIN_VALUE;
+        //SudokuBoard maxBoard = null;
+        for (int i = 1; i <= numToCreate; i++){
+            long time = System.currentTimeMillis();
+            SudokuBoard board = generator.getProblem();
+            time = System.currentTimeMillis() - time;
+
+            int numClues = board.getNumClues();
+            int numCells = board.getNumCells();
+            if (numClues < minClues){
+                minClues = numClues;
+                minBoard = board;
+            }
+            /*
+            if (time > maxTime){
+                maxTime = time;
+                maxBoard = board;
+            }
+                    */
+            totalClues += numClues;
+            totalTime += time;
+
+            System.out.print("problem " + i + " of " + numToCreate + ": ");
+            System.out.println(numClues + " / " + numCells + " problem created, took " + time + "ms");
+            if (verbose){
+                System.out.println(board);
+                System.out.println();
+            }
+            problems.add(board);
+        }
+
+        System.out.println("\nboard with minimal number of clues ("+minClues+") created:\n" + minBoard);
+        //System.out.println("\nboard that took longest to create ("+maxTime+"ms):\n" + maxBoard);
+
+        System.out.println("\nsolver statistics");
+        System.out.println("        strategy        |  wins  ");
+        System.out.println("------------------------+--------");
+        Map<Class,Integer> wins = SolverService.getWinMap();
+        for (Class strat : wins.keySet()){
+            String name = " " + StringUtil.padRight(strat.getSimpleName(), ' ', 23);
+            System.out.println(name + "|  " + wins.get(strat));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.